packages/dev/loaders/src/USD/usdFileLoader.ts

```diff
@@ -0,0 +1,242 @@
+import { Tools } from "core/Misc/tools";
+import { Constants } from "core/Engines/constants";
+import type { USDLoadingOptions } from "./usdLoadingOptions";
+
+declare module "core/Loading/sceneLoader" {
+    // eslint-disable-next-line jsdoc/require-jsdoc
+    export interface SceneLoaderPluginOptions {
+        /**
+         * Defines options for the usd loader.
+         */
+        [USDFileLoaderMetadata.name]: Partial<USDLoadingOptions>;
+    }
+}
+
+/**
+ * @experimental
+ * USD(z) file type loader.
+ * This is a babylon scene loader plugin.
+ */
+export class USDFileLoader implements ISceneLoaderPluginAsync, ISceneLoaderPluginFactory {
+    /**
+     * Defines the name of the plugin.
+     */
+    public readonly name = USDFileLoaderMetadata.name;
+
+    private _assetContainer: Nullable<AssetContainer> = null;
+
+    private readonly _loadingOptions: Readonly<USDLoadingOptions>;
+    /**
+     * Defines the extensions the UDS loader is able to load.
+     * force data to come in as an ArrayBuffer
+     */
+    public readonly extensions = USDFileLoaderMetadata.extensions;
+
+    /**
+     * Creates loader for USD files
+     * @param loadingOptions options for loading and parsing usdz files.
+     */
+    constructor(loadingOptions: Partial<Readonly<USDLoadingOptions>> = USDFileLoader._DefaultLoadingOptions) {
+        this._loadingOptions = loadingOptions;
+    }
+
+    /** @internal */
+    createPlugin(): ISceneLoaderPluginAsync {
+        return new USDFileLoader();
+    }
+
+    private static readonly _DefaultLoadingOptions = {
+        usdLoaderUrl: "https://lighttransport.github.io/tinyusdz/",
+    } as const satisfies USDLoadingOptions;
+
+    /**
+     * Imports from the loaded USD data and adds them to the scene
+     * @param meshesNames a string or array of strings of the mesh names that should be loaded from the file
+     * @param scene the scene the meshes should be added to
+     * @param data the USD data to load
+     * @param rootUrl root url to load from
+     * @param onProgress callback called while file is loading
+     * @param fileName Defines the name of the file to load
+     * @returns a promise containing the loaded meshes, particles, skeletons and animations
+     */
+    public async importMeshAsync(
+        meshesNames: any,
+        scene: Scene,
+        data: any,
+        rootUrl: string,
+        onProgress?: (event: ISceneLoaderProgressEvent) => void,
+        fileName?: string
+    ): Promise<ISceneLoaderAsyncResult> {
+        return this._parse(meshesNames, scene, data, rootUrl).then((meshes) => {
+            return {
+                meshes: meshes,
+                particleSystems: [],
+                skeletons: [],
+                animationGroups: [],
+                transformNodes: [],
+                geometries: [],
+                lights: [],
+                spriteManagers: [],
+            };
+        });
+    }
+    private static _UniqueResolveID = 1000;

```

Comment:
could be module-level function (instead of private static)

Reply:
Moved into its own private script